### PR TITLE
refactor: remove a bunch of closure allocation

### PR DESCRIPTION
**Pull request by rgrinberg**

### CODE REVIEW

File: lib/hamt.ml [hunk 1]
```diff
@@ -231,36 +230,43 @@ module Make (Config : CONFIG) (Key : Hashtbl.HashedType) :
...
-        else x :: update_list update k xs
+  let update_list =
+    let rec loop update k = function
+      | [] -> ( match update None with None -> [] | Some v -> [ (k, v) ])
+      | ((kx, vx) as x) :: xs ->
+          if kx = k then
+            match update (Some vx) with None -> xs | Some v -> (k, v) :: xs
+          else x :: loop update k xs
+    in
+    fun update k xs -> loop update k xs
```
Comment (thizanne) [inline lib/hamt.ml:241]: I don't get which closure you're removing here, apart from inling `option`?

**Review by thizanne — APPROVED** [review]

File: lib/hamt.ml (hunk 1 above)
Comment (rgrinberg) [reply lib/hamt.ml:241]: `(fun v -> (k, v) :: xs)` is the closure needed for `option`.

Comment (rgrinberg) [reply lib/hamt.ml:241]: But you're right, I don't know why I added this intermediate `loop`. I'll get rid of it.

Comment (rgrinberg) [reply lib/hamt.ml:241]: I'll fix in a subsequent PR. If i fix it here, I will create a conflict for myself.

**Pull request merged**